Source/monster.cpp

```diff
@@ -4653,56 +4653,40 @@ bool IsGoat(_monster_id mt)
 	    MT_NGOATBW, MT_BGOATBW, MT_RGOATBW, MT_GGOATBW);
 }
 
-bool SpawnSkeleton(Monster *monster, Point position)
+void ActivateSkeleton(Monster &monster, Point position)
 {
-	if (monster == nullptr)
-		return false;
-
 	if (IsTileAvailable(position)) {
-		Direction dir = GetDirection(position, position); // TODO useless calculation
-		ActivateSpawn(*monster, position, dir);
-		return true;
+		ActivateSpawn(monster, position, Direction::SouthWest);
+		return;
 	}
 
-	bool monstok[3][3];
+	constexpr std::array<Direction, 8> spawnDirections {
+		Direction::North, Direction::NorthEast, Direction::East, Direction::NorthWest, Direction::SouthEast, Direction::West, Direction::SouthWest, Direction::South
+	};
+	std::bitset<8> spawnOk;
 
-	bool savail = false;
-	int yy = 0;
-	for (int j = position.y - 1; j <= position.y + 1; j++) {
-		int xx = 0;
-		for (int k = position.x - 1; k <= position.x + 1; k++) {
-			monstok[xx][yy] = IsTileAvailable({ k, j });
-			savail = savail || monstok[xx][yy];
-			xx++;
-		}
-		yy++;
-	}
-	if (!savail) {
-		return false;
+	for (size_t i = 0; i < spawnDirections.size(); i++) {
+		if (IsTileAvailable(position + spawnDirections[i]))
+			spawnOk.set(i);
 	}
+	if (spawnOk.none())
+		return;
 
-	int rs = GenerateRnd(15) + 1;
-	int x2 = 0;
-	int y2 = 0;
-	while (rs > 0) {
-		if (monstok[x2][y2])
-			rs--;
-		if (rs > 0) {
-			x2++;
-			if (x2 == 3) {
-				x2 = 0;
-				y2++;
-				if (y2 == 3)
-					y2 = 0;
-			}
-		}
-	}
+	// this is used in the following loop to find the nth set bit.
+	int spawnChoice = GenerateRnd(15) % spawnOk.count();

```

Comment:
Maybe we could also use count to clean up https://github.com/diasurgical/devilutionX/blob/master/Source/levels/drlg_l1.cpp#L545 and it's siblings